RAG Optimization, step 1 (#132)
aiken-lang/site: src/pages/faq.md

**Key Points:**
- Command: `aiken new foo/bar`
- Creates `aiken.toml`, `lib/`, `validators/`, and `README.md`
- The standard library is automatically added as a dependency
- Library code goes in `lib/`; validators go in `validators/`

**Example:**
```bash
aiken new aiken-lang/hello-world
cd hello-world
```

Resulting structure:
```
.
├── README.md
├── aiken.toml
├── lib
│   └── hello-world
└── validators
```

**Related:** [Project Structure](#q6-what-is-the-project-structure-of-an-aiken-project) | [Compiling](#q7-how-do-i-compile-an-aiken-project)

---

## Q6: What is the project structure of an Aiken project?

**Answer:** Aiken projects split source code into two categories: library code (in `lib/`) and application code, i.e. on-chain validators (in `validators/`). The project root contains an `aiken.toml` configuration file. After compilation, a Plutus blueprint (`plutus.json`) is generated containing compiled validator code and hash digests.

**Key Points:**
- `lib/` — Reusable library modules

> - `validators/` — On-chain validator source files (`.ak`)

Reviewer: ```suggestion
- `validators/` — On-chain validator source files (`.ak`)
- `env/` — Environment-specific code that can be conditionally included  
```

- `aiken.toml` — Project metadata, dependencies, and configuration
- `plutus.json` — Generated Plutus blueprint (CIP-0057) after build
- `build/` — Build artifacts and fetched dependencies

**Example (`aiken.toml`):**
```toml
name = "foo/bar"
version = "1.0.0"
licence = "Apache-2.0"
description = "A next-level DeFi platform"

[[dependencies]]
source = "github"
name = "aiken-lang/stdlib"
version = "main"
```

**Related:** [Creating a Project](#q5-how-do-i-create-a-new-aiken-project) | [Compiling](#q7-how-do-i-compile-an-aiken-project)

---

## Q7: How do I compile an Aiken project?

**Answer:** Use `aiken build` to compile your project, which generates the Plutus blueprint (`plutus.json`). Use `aiken check` to type-check and run tests without producing a full build. For library projects, `aiken docs` generates HTML documentation from types, annotations, and comments.

**Key Points:**
- `aiken build` — Full compile, generates `plutus.json`

> - `aiken check` — Type-check + run tests only

Reviewer: ```suggestion
- `aiken check` — Type-check + run tests only
- `aiken bench` — Type-check + run benchmarks only
```

- `aiken docs` — Generate HTML documentation
- `aiken blueprint` — Generate addresses, apply parameters, convert formats
- Traces are removed by default on `aiken build` (use `--trace-level verbose` to keep them)
- Traces are kept by default on `aiken check`

**Related:** [Testing](#q17-how-do-i-test-validators-in-aiken) | [Troubleshooting](#q24-how-do-i-debug-and-troubleshoot-validators)

---

## Q8: What is the eUTxO model?

**Answer:** The Extended Unspent Transaction Output (eUTxO) model is Cardano's transaction model. In this model, transactions consume existing UTxOs (inputs) and produce new UTxOs (outputs). Each output has a value (assets) and an address (spending conditions). The "extended" part adds datums and redeemers, enabling stateful smart contract logic while maintaining full determinism.

**Key Points:**
- A UTxO (Unspent Transaction Output) has a value and an address
- Inputs reference previous outputs; spent outputs are destroyed
- Transactions are atomic: all-or-nothing
- The eUTxO extension adds datums (state/config attached to outputs) and redeemers (user arguments provided at spending time)

> - Scripts (validators) are deterministic predicates — they return True or False

Reviewer: ```suggestion
- Scripts (validators) are akin to deterministic predicates — conceptually, they return True or False
```

**Key Points:**
- Unit tests: functions with no arguments returning `Bool`; use the `test` keyword
- Property-based tests: take a fuzzer argument via the `via` keyword
- Run tests with `aiken check`
- Tests collect traces for debugging
- Automatic diffing on test failures shows what went wrong
- Test failures using `fail` keyword test expected-failure paths
- Filter tests with `-m` flag: `aiken check -m "module_name"`
- Tests run on the same virtual machine as production validators

**Example:**
```aiken
// Unit test
test simple_addition() {
  let result = 2 + 3
  result == 5
}

// Property-based test
use aiken/fuzz

test prop_commutative((a, b) via fuzz.both(fuzz.int(), fuzz.int())) {
  a + b == b + a
}

// Expected failure test
test must_fail() fail {
  expect Some(result) = math.sqrt(-42)
  result == -1
}
```

**Related:** [Property-Based Testing](#q18-what-is-property-based-testing-in-aiken) | [Troubleshooting](#q24-how-do-i-debug-and-troubleshoot-validators)

---

## Q18: What is property-based testing in Aiken?

> **Answer:** Property-based testing is a first-class feature in Aiken that generates random test inputs to check general properties rather than specific cases. It includes integrated shrinking, which automatically simplifies counterexamples to the smallest failing input. Fuzzers are introduced using the `via` keyword and the `aiken/fuzz` library provides composable fuzzer primitives.

Reviewer: ```suggestion
**Answer:** Property-based testing is a first-class feature in Aiken that generates random test inputs to check general properties rather than specific cases. It includes integrated shrinking, which automatically simplifies counterexamples to smaller failing inputs. Fuzzers are introduced using the `via` keyword and the `aiken/fuzz` library provides composable fuzzer primitives.
```

---

## Q20: How does pattern matching work in Aiken?

**Answer:** Aiken uses `when/is` expressions for pattern matching, similar to `match` in Rust or `case` in Haskell. Pattern matching is exhaustive — the compiler checks that all cases are covered. You can destructure custom types, lists, tuples, and more directly in patterns.

**Key Points:**
- `when value is { pattern -> result }` for multi-case matching
- Patterns can destructure lists: `[]`, `[x]`, `[x, y, ..]`
- Patterns can destructure custom types by constructor
- Wildcard `_` matches anything
- Must be exhaustive — compiler enforces coverage of all cases
- `expect` is used for non-exhaustive matching (fails if pattern doesn't match)

**Example:**
```aiken
fn describe_list(list: List<Int>) -> String {
  when list is {
    [] -> @"The list is empty"
    [x] -> @"The list has one element"
    [x, y, ..] -> @"The list has multiple elements"
  }
}

fn unwrap_option(opt: Option<Int>) -> Int {
  when opt is {
    Some(a) -> a
    None -> fail @"Expected Some but got None"
  }
}
```

**Related:** [Custom Types](#q15-how-do-i-define-custom-types) | [Error Handling](#q21-how-do-i-handle-errors-in-aiken)

---

> ## Q21: How do I handle errors in Aiken?

Reviewer: ```suggestion
## Q21: How do I deal with errors in Aiken?
```
